### PR TITLE
Progressive Multi-Jittered Sample Sequences

### PR DESCRIPTION
**Pull request by Alpistinho**
Hi,

I have been working on porting [this patch](https://developer.blender.org/D4686) from Cycles to LuxCoreRender. The point generation is exactly the same code, only adapted to fit in the different structure.

I think I've finally succeeded, but unfortunately the performance (at equals samples/pixel) is not better than Sobol provides currently.

Their implementation is very close to the [reference code ](https://graphics.pixar.com/library/ProgressiveMultiJitteredSampling) provided by the authors.

At 16 Spp
Sobol:
normal
 
PMJ02 (with shuffling):
normal_pmj02_16spp1_thread

PMJ02 (without shuffling):
normal_pmj02_16spp

As you can see, without using shuffling the performance is clearly worse. However, it seems to me that without shuffling it is improved if we discount the heavy sample correlation. (Can we do that?)

I've produced some point distribution graphs comparing the shuffling across dimensions and while each 2D table of samples seem to be preserved, the distribution between two different tables is less than ideal

First two dimensions (No shuffling)
dim_0_1_no_shuffling
First and third dimensions (No shuffling)
dim_0_2_no_shuffling
First two dimensions (Shuffled)
dim_0_1_shuffling
First and third dimensions (Shuffled)
dim_0_2_shuffling

Does anyone have any suggestion?

The paper states:

> It is fairly easy to generalize the pmj sample generation algorithms to three dimensions. For four dimensions or higher, we com-bine 2D pmj02 sequences by locally shuffling the samples in pairs of dimensions. Typically such shuffling is done just to decorrelatethe dimensions [Coo86], but since we use tables we can do better:shuffle such that the samples are stratified in 4D (while maintainingthe full (0,2) property in each pair of dimensions). This shuffling is implemented as a simple “greedy” algorithm: for all pairs of 2Dpoints within segments of a (truncated) sequence, swap the points if swapping them reduces the number of 4D strata with more than one sample. This results in a sequence with full 4D stratification(for example 4×4×2×2 strata for the first 64 samples) and takesonly a few minutes for two 2D sequences with 4096 samples.

I don't really know how to do that, and it seems to be a lenghty process, so we would have to generate the samples beforehand.

Other know issues:
* Currently each thread generates it's own sequences. They should be shared accross thread
* If this is made to work, the GPU would need to load that data, since the code will not be fast enough to be run there.

### CODE REVIEW

**Comment by Dade916** [comment]
@Alpistinho, I think you would need to identify 1D (which light source to sample, etc.), 2D (camera lens coordinates, BSDf sampling, etc.), etc. samples inside the generic samples array to have this kind of stuff to work.

This is not currently part of the Sampler interface and it can not be because different render engines will use samples in different ways and with different coupling.
You can only handle the samples array as a n-dimensional vector; 1D, 2D, etc. samples do not exist in LuxCore.

Short version: the original paper idea is not applicable in our context.

P.S. people have annoyed me for years because of the samples correlation in original Sobol code.

**Comment by Alpistinho** [comment]
Ok then.

Do you know if cycles implement this concept?

I imagine implementing this here would be a huge amount of work, right?

**Comment by Dade916** [comment]
I don't know, may be, Cycles is a single type of render engine so it may be possible to know in what context each sample is used.

**Comment by Alpistinho** [comment]
Do you think it would be worth investing time to introduce this kind of concepts or should we just close this PR?

I guess that would be quite a change, right?

**Comment by Alpistinho** [comment]
Hi,

Taking from the issue you've pointed about the the lack of 1D and 2D sample concepts in Lux, I've extended the Sampler interface to include this possibility and adapted the PathCPU to use these samples.

Sobol (64 spp/240s):
normal_sobol_64spp_new_240s

PMJ02 (64 spp/215s)
normal_pmj02_64spp_new_215s_dimensionShuffling

Now I think PMJ02 is winning now, right? =)
(I forgot to create a render with the current Sobol, this one is already using the 1D/2D interface and that changes the dimension ordering, need to check if that has any effect for the Sobol sampler)

Do you think it is worth bringing the 1D/2D sample concept to the other Engines and maybe to other classes that also deal with these samples? 

Current known issues:
* Still a sequence per thread. Using more than one thread worsens the PMJ02 sampling quality since every thread would be using a different sequence. I intend to move the sample generation to the SharedSampler and each Sampler instance load a copy of the data
* I switched from scrambling the pass to scrambling the dimension. The result seems to be better that way, but there is some temporal flickering at the beggining. Maybe generating more dimensions than strictly needed would alleviate the problem. 
  * I think that even though the PMJ02 sequence can be infinitely generated keeping its properties, I cannot start it from an arbitrary point and expect the same, unless I intend to cover it entirely anyways.
* Obviously, I haven't tested that very extensively.

**Comment by Dade916** [comment]
Can you try with some "real" scene ? Like a couple of the one included in https://github.com/LuxCoreRender/LuxCoreTestScenes

At the moment, the improvement looks very small to me and hardly justify the amount of work required to update all sampler, render engines and all OpenCL code.

**Comment by Alpistinho** [comment]
Some testing:

All of them show somewhat lower noise for the PMJ02 sampler.
What do you think of it?

I have already ported all Samplers to support the new interface and wouldn't mind doing that for the rest of the engines, bsdf, light sampling and everything else. I know it would be a longer term project than just a few days, but I think it is doable if I just work on a separate branch. It is basically changing some interfaces to match each other all around.

There might be still things to improve for the current code, so maybe some performance improvement is still possible. I thought that it might be worth using Sobol for the 1D sequences, since the distribution between different 2D sequences is not good. 
Another improvement that I though about is that instead of just rotating which dimension each pixel uses using the same worded, I could shuffle it completely from pixel to pixel.

In any case, I understand if you think this would be too much work for the benefit, so we can stop this here if it is not worth it.

## Food Scene

### Sobol 64 Spp
food_sobol_64spp_600px

### PMJ02 64 Spp
food_pmj02_64spp

## ProceduraLeaves

### Sobol 16 Spp
proceduraleave_sobol_16spp

### PMJ02 16 Spp
proceduraleave_pmj02_16spp

## LuxCore Bench 2.1 

### Sobol 32 Spp
luxbench_sobol_32spp

### PMJ02 32 Spp
luxbench_pmj02_32spp

**Comment by Alpistinho** [comment]
Just tried both things I suggested above:
* Shuffling the dimension order from pixel to pixel seem to remove the flickering completely, so that's good
* Using Sobol to generate the 1D dimensions makes so little difference that I can't tell if it is better or not.
  * The sample correlation is probably not a big factor if they are not actually being used to sample the same thing

**Comment by Alpistinho** [comment]
Not that much difference, it seems

### Sobol 128spp with adaptive sampling
food_128spp_sobol
### PMJ02 128spp with adaptive sampling
food_pmj02_128spp_complete_shuffling

**Comment by Alpistinho** [comment]
Let's leave it here like a proof of concept then.
